test/fork/MigrationActions.t.sol

```diff
@@ -0,0 +1,522 @@
+        public assertDebtStateDoesNotChange
+    {
+        amount = _bound(amount, 0, DAI_SUPPLY);
+
+        deal(DAI, user, amount);  // Use `deal` to get a higher DAI amount
+
+        _runMigrateDAIToNSTTest(amount);
+    }
+
+}
+
+contract MigrateDaiToSNstIntegrationTest is MigrationActionsIntegrationTestBase {
+
+    // Starting balance of NST in the SNST contract
+    uint256 startingBalance = 1051.297887154176590368e18;
+
+    function _runMigrateDAIToSNSTTest(uint256 amount) internal {
+        // Get the expected amount to be sucked from the vat on `drip` in deposit call in sNST
+        uint256 snapshot   = vm.snapshot();
+        uint256 nstBalance = nst.balanceOf(SNST);
+        snst.drip();
+        uint256 nstDripAmount = nst.balanceOf(SNST) - nstBalance;
+        vm.revertTo(snapshot);
+
+        dai.approve(address(actions), amount);
+
+        assertEq(dai.balanceOf(user), amount);
+        assertEq(nst.balanceOf(SNST), startingBalance);
+
+        assertEq(snst.convertToAssets(snst.balanceOf(user)), 0);
+
+        uint256 debt      = vat.debt();
+        uint256 sumSupply = _getSumSupply();
+
+        actions.migrateDAIToSNST(user, amount);
+
+        assertEq(dai.balanceOf(user), 0);
+        assertEq(nst.balanceOf(SNST), startingBalance + nstDripAmount + amount);
+
+        // Assert within 2 wei diff, rounding down
+        assertLe(amount - snst.convertToAssets(snst.balanceOf(user)), 2);
+
+        assertEq(vat.debt(), debt + nstDripAmount * 1e27);
+
+        // Two rounding events in nst.totalAssets()
+        assertApproxEqAbs(_getSumSupply(), sumSupply, 2);
+    }
+
+    function test_migrateDAIToSNST() public {
+        uint256 amount = 1000 ether;
+
+        _getDai(user, amount);
+
+        _runMigrateDAIToSNSTTest(amount);
+    }
+
+    function testFuzz_migrateDAIToSNST(uint256 amount) public {
+        amount = _bound(amount, 0, dai.balanceOf(DAI_WHALE));
+
+        _getDai(user, amount);
+
+        _runMigrateDAIToSNSTTest(amount);
+    }
+
+    function testFuzz_migrateDAIToSNST_upToWholeSupply(uint256 amount) public {
+        amount = _bound(amount, 0, DAI_SUPPLY);
+
+        deal(DAI, user, amount);
+
+        _runMigrateDAIToSNSTTest(amount);
+    }
+
+}
+
+contract MigrateSDaiAssetsToNstIntegrationTest is MigrationActionsIntegrationTestBase {
+
+    function _runMigrateSDAIAssetsToNSTTest(uint256 amount) internal {
+        // Deposit into sDAI
+        dai.approve(SDAI, amount);
+        sdai.deposit(amount, address(this));
+
+        // Warp to accrue value in both sDAI and sNST after drip is called on sDAI deposit
+        skip(2 hours);
+
+        // Get the expected amount to be sucked from the vat on `drip` in withdraw call in sDAI
+        uint256 snapshot      = vm.snapshot();
+        uint256 preDripPotDai = vat.dai(POT);
+        pot.drip();
+        uint256 daiDripAmount = vat.dai(POT) - preDripPotDai;
+        vm.revertTo(snapshot);
+
+        sdai.approve(address(actions), amount);
```

Comment:
The approval is for the shares amount, so this can be tighter.

Reply:
Updated